packages/server/lib/controllers/auth/postTwoStep.ts

```diff
@@ -0,0 +1,216 @@
+            error: { code: 'invalid_uri_params', errors: zodErrorToHTTP(paramsVal.error) }
+        });
+        return;
+    }
+
+    const { account, environment, authType } = res.locals;
+    const bodyData: PostPublicTwoStepAuthorization['Body'] = val.data;
+    const { connection_id: receivedConnectionId, params, hmac }: PostPublicTwoStepAuthorization['Querystring'] = queryStringVal.data;
+    const { providerConfigKey }: PostPublicTwoStepAuthorization['Params'] = paramsVal.data;
+    const connectionConfig = params ? getConnectionConfig(params) : {};
+
+    let logCtx: LogContext | undefined;
+
+    try {
+        logCtx = await logContextGetter.create(
+            {
+                operation: { type: 'auth', action: 'create_connection' },
+                meta: { authType: 'twostep' },
+                expiresAt: defaultOperationExpiration.auth()
+            },
+            { account, environment }
+        );
+        void analytics.track(AnalyticsTypes.PRE_TWO_STEP_AUTH, account.id);
+
+        await hmacCheck({
+            environment,
+            logCtx,
+            providerConfigKey,
+            connectionId: receivedConnectionId,
+            hmac,
+            res
+        });
+
+        const config = await configService.getProviderConfig(providerConfigKey, environment.id);
+        if (!config) {
+            await logCtx.error('Unknown provider config');
+            await logCtx.failed();
+            res.status(404).send({ error: { code: 'unknown_provider_config' } });
+            return;
+        }
+
+        const provider = getProvider(config.provider);
+        if (!provider) {
+            await logCtx.error('Unknown provider');
+            await logCtx.failed();
+            res.status(404).send({ error: { code: 'unknown_provider_template' } });
+            return;
+        }
+
+        if (provider.auth_mode !== 'TWOSTEP') {
+            await logCtx.error('Provider does not support TWOSTEP auth', { provider: config.provider });
+            await logCtx.failed();
+            res.status(400).send({ error: { code: 'invalid_auth_mode' } });
+            return;
+        }
+
+        await logCtx.enrichOperation({ integrationId: config.id!, integrationName: config.unique_key, providerName: config.provider });
+
+        const {
+            success,
+            error,
+            response: credentials
+        } = await connectionService.getTwoStepCredentials(provider as ProviderTwoStep, bodyData, connectionConfig);
+
+        if (!success || !credentials) {
+            await logCtx.error('Error during TwoStep credentials creation', { error, provider: config.provider });
+            await logCtx.failed();
+
+            errorManager.errRes(res, 'two_step_error');
+
+            return;
+        }
+
+        const connectionId = receivedConnectionId || connectionService.generateConnectionId();
+
+        const [updatedConnection] = await connectionService.upsetTwoStepConnection({
```

Comment:
typo: `upsert` instead of `upset`

Reply:
updated with the latest commit, thank you 🙏!